docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+most immediately, these will be the rebalance system and the repair system.
+These systems, when identifying a relocation opportunity, will simply perform
+the first operation of the relocation, and the replicate queue will perform the
+complementary operation. 
+
+RFCs for these operations are forthcoming.
+
+The final complication is how to ensure that the replicate queue promptly
+identifies ranges outside of their ideal target state. As a queue it will be
+attached to the replica scanner, but we will also want to enqueue a replica
+immediately whenever we knowingly perturb the replication state. Thus,
+components initating a relocation (e.g. rebalance or repair) should immediately
+enqueue their target replicas after changing the replication state. 
+
+# Drawbacks
+
+### Specific Move Operations
+A stateless model for relocation precludes the ability to request specific
+relocations; only the first operation can be made with specificity.
+
+For example, the verb "Move replica from X to Y" cannot be expressed with
+certainty; instead, only "Move replica to X from (some store)" or "Move replica
+to Y from (some store)" can be expressed. The replicate queue will be responsible
+for selecting an appropriate store for the complementary operation.
+
+It is assumed that this level of specificity is simply not needed for any
+relocation operations; there is no currently apparent use case where a move
+between a specific pair of stores is needed.
+
+Even if this was necessary, it might be possible to express by manipulating the
+factors behind the individual stateless decisions.
+
+### Thrashing of complementary operation
+Because there is no relocation state, the possibilty of "thrashing" is introduced. For
+example: 
+
+1. Rebalance operation adds a new replica to the range.
+2. The replicate queue picks up the range and detects the need to remove a
+   replica; however, it decides to remove the replica that was just added.
+
+This is possible if the rebalance's criteria for new replica selection are
+sufficiently different from the replicate queue's selection criteria for
+removing a replica.
+
+To reduce this risk, there must be sufficient agreement in the criteria between
+the operations; a Rebalance operation should avoid adding a new replica if
+there's a realistic chance that the replicate queue will immediately remove it.
+
+This can realized by adding a "buffer" zone between the different criteria; that
+is, when selecting a replica to add, the new replica should be *significantly*
+removed from the criteria for removing a replica, thus reducing the chances that
+it will be selected.
+
+When properly tuned to avoid thrashing, the stateless nature could instead be
+considered a positive because it can respond to changes during the interim;
+consider if, due to a delay, the relative health of nodes changes and the
+originally selected replica is no longer a good option.  The stateless system
+will respond correctly to this situation.
```

Comment:
I agree. You want to place a new replica on a node that has low load, and want to remove a replica from a node that has high load. As load changes you can easily thrash, but that's how the system should behave.